fix(@mastra/core): apply output processors before saving to memory

TylerBarnes opened the pull request:
Fixes #7087

Output processors were modifying the streamed response but the original unmodified text was still being saved to memory. This was particularly problematic for PII redaction use cases.

The fix moves output processor application earlier in the stream pipeline, ensuring transformations happen before messages are persisted to storage.

Before this fix, if you had a processor that redacted sensitive data:
```typescript
class RedactionProcessor implements Processor {
  async processOutputStream({ part }) {
    if (part.type === 'text-delta' && part.text) {
      return { ...part, text: part.text.replace(/SSN-\d{3}-\d{2}-\d{4}/g, '[REDACTED]') };
    }
    return part;
  }
}
```

The user would see `[REDACTED]` in the stream, but refreshing the page would show the original SSN from memory.

Now the processor runs before memory storage, so both the stream and stored messages show `[REDACTED]` consistently.

TylerBarnes committed:
test: add failing test demonstrating stream vs non-stream processor disconnect

Reproduces issue #7087 where processOutputStream modifications are not
reflected in processOutputResult - the stream shows redacted content
but processOutputResult receives original unprocessed messages

test

TylerBarnes committed:
summary

TylerBarnes committed:
fix(core): ensure output processors transform both stream and final messages

Fixes #7087 where stream processors (processOutputStream) were not affecting
the final messages passed to processOutputResult.

Changes:
- Pass outputProcessors from Agent to LLM stream methods
- Remove duplicate stream processing that was causing text duplication
- Update messageList with processed text before calling output processors
- Add tests to verify consistency between streamed and final output

The fix ensures that when processors transform streaming text (e.g., redacting
sensitive data), those transformations are preserved in the final messages
passed to processOutputResult.

TylerBarnes committed:
Revert "summary"

This reverts commit d37fd5c45fd36fd6e38af9c59f00a8ecf6576b11.

TylerBarnes committed:
revert incorrect fix

TylerBarnes committed:
fix(core): apply output processors before saving to memory (#7087)

- Move output processor application earlier in the stream pipeline
- Ensure processors run before messages are saved to memory storage
- Add test demonstrating processor consistency between stream and memory
- Prevent double processing by applying processors only in baseStream

This fixes the issue where output processors (like PII redaction) would
modify the streamed response but the original unmodified text would
still be saved to the database.

DanielSLew commented on packages/core/src/agent/index.ts:
```diff
@@ -279,6 +279,22 @@ export class Agent<
+    }
+
+    if (typeof this.#outputProcessors === 'function') {
+      if (!runtimeContext) {
+        // Create a default runtime context if none provided
+        return [];
+      }
```
The comment contradicts what the code is doing, don't we want to return the output processors even if there's no runtime context

DanielSLew commented on packages/core/src/stream/base/output.ts:
```diff
@@ -560,41 +602,10 @@ export class MastraModelOutput<OUTPUT extends OutputSchema = undefined> extends
     let fullStream = this.teeStream();
 
-    const processorStates = new Map<string, ProcessorState>();
+    // Output processors are now applied in the baseStream transform
+    // so we don't need to apply them here anymore
```
```suggestion
```

TylerBarnes committed:
Update packages/core/src/stream/base/output.ts

Co-authored-by: Daniel Lew <51924260+DanielSLew@users.noreply.github.com>

TylerBarnes committed:
fix: create default RuntimeContext when none provided

Addresses PR comment: https://github.com/mastra-ai/mastra/pull/7647#discussion_r2334682622
The comment now correctly reflects the code behavior - we create a default RuntimeContext
when none is provided for function-based output processors.

TylerBarnes committed:
prettier

TylerBarnes committed:
fix: resolve text promise when tripwire is caught

- Ensure text promise is resolved with empty string when TripWire is caught
- Fix unused parameter warning in ProcessorState constructor
- This partially addresses test failures where processOutputResult needs to abort

TylerBarnes committed:
fix tripwire

TylerBarnes committed:
Merge branch 'main' into fix/issue-7087

TylerBarnes committed:
run stream processing on inner stream and result processing on outer stream
